addons/account/models/account_tax.py

```diff
@@ -1157,21 +1157,14 @@ def grouping_key_generator(base_line, tax_values):
         amount_tax = 0.0
 
         subtotal_order = {}
-        groups_by_subtotal = {}
+        groups_by_subtotal = defaultdict(list)
         for tax_group_vals in tax_group_vals_list:
             tax_group = tax_group_vals['tax_group']
 
-            if tax_group.preceding_subtotal:
-                subtotal_title = tax_group.preceding_subtotal
-                sequence = tax_group.sequence + 1 # Avoid sequence = 0 here.
-            else:
-                subtotal_title = _("Untaxed Amount")
-                sequence = 0
-
-            if subtotal_title not in subtotal_order:
-                subtotal_order[subtotal_title] = sequence
-                groups_by_subtotal[subtotal_title] = []
+            subtotal_title = tax_group.preceding_subtotal or _("Untaxed Amount")
+            sequence = tax_group.sequence
 
+            subtotal_order[subtotal_title] = min(subtotal_order.get(subtotal_title, float('inf')), sequence)
```

Comment:
This refactoring is not fixing anything, right?

Reply:
Oh, but it does.
If you look at the commit message

> This also allows more flexibility for ordering tax groups.
> Before https://github.com/odoo-dev/odoo/commit/cd6c33575b4823ebf6ab7100552c668360f6b30d, it was possible to have this:
> * Tax Exempt 2% (preceeding subtotal "Discounted")
> * VAT 21% (no preceeding subtotal, take default)
> * Reapply 2% (preceeding subtotal "Reapply")
> 
> The value for groups without preceeding subtotals was always 0 and
> * before the commit mentionned above, it was possible to reach 0 too for
>   the groups with a preceeding subtotal
> * but after +1 was added making it impossible from the view
> 
> We now always take the group's order (or the smallest one for groups
> without a preceeding subtotal)

Reply:
You should add a test for that then (in `test_invoice_tax_totals.py`)

Reply:
I have le feu au cul :weary:

Reply:
Done